GitCommands/ArgumentBuilder.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Collections;
+using System.Text;
+using JetBrains.Annotations;
+
+namespace GitCommands
+{
+    /// <summary>
+    /// Builds a command line argument string from zero or more arguments.
+    /// </summary>
+    /// <remarks>
+    /// To retrieve the constructed argument list string, call <see cref="ToString"/>.
+    /// <para />
+    /// Arguments are separated by a single space character.
+    /// <para />
+    /// Adding <c>null</c> or white-space strings has no effect on the output, which can be
+    /// useful in some calling constructions.
+    /// <para />
+    /// This class has been designed to work with C# collection initialiser syntax which makes
+    /// its use quite ergonomic. See the example for more information.
+    /// <para />
+    /// The type accepts strings, however conversion from other types (eg. <see cref="ForcePushOptions"/>
+    /// enum) is achieved via extension methods on <see cref="ArgumentBuilderExtensions"/> by adding a
+    /// method named <c>Add</c> that accepts the required type.
+    /// </remarks>
+    /// <example>
+    /// <code>
+    /// var args = new ArgumentBuilder
+    /// {
+    ///     "commit",                   // adds the string unnconditionally
+    ///     { isAmend, "--amend" },     // adds the option only if isAmend == true
+    ///     { isUp, "--up", "--down" }, // selects the option based on the value of isUp
+    /// };
+    /// </code>
+    /// </example>
+    public sealed class ArgumentBuilder : IEnumerable
```

Comment:
💭 This could probably be a `struct` to avoid the extra allocations.

Reply:
Good idea. Done.

Reply:
I had to back this out as it broke logic. The extension methods accept `ArgumentBuilder` as a parameter, which ends up being a copy. The `this` param cannot be `ref` unfortunately. That'd be useful for structs.

This is marked as intended for 7.2, but that ship has sailed. Am I missing something?

https://github.com/dotnet/csharplang/issues/186

Reply:
What error were you seeing? For this type, even copies should behave the same as the original.